### DISCUSSION TITLE
About System.Range Extension Support

### DISCUSSION THREAD

**Discussion by yingpanwang** [dotnet/csharplang#8267] [Q&A] [answered]
I wrote an extension method for Range, but I encountered a problem that Index is non-negative

Currently, only Range supports the range expression Start...End.

There is no room for extensions for other classes with the same Range semantics.

Extract an abstract interface for Range.

The compiler-generated code for Range should be changed to use this interface.

If possible, it could even be 

```csharp

IRange range = new MyRange('a', 'c');

foreach(var item in range)
   ......

foreach(var item in a...c)
  ......

```

Of course, the std library only needs to generate a default Range when a specific implementation is not specified.

### Additional context

![image](https://github.com/dotnet/aspnetcore/assets/49632937/91f84374-4c52-4b83-904b-a24ec69d46e6)

**Comment by huoyaoyuan** [comment]
> with the same Range semantics

Being non-negative for `Index` is a part of the semantics of `Range`.

For such `IEnumerable` extension, it makes even more sense to use `(int, int)` instead.

**Reply by yingpanwang** [reply]
What I mean is that the semantics of Range are limited by the current implementation of the Index struct.

Instead of the usual semantics of Range representing an interval.

Is it possible to extend the current Range, for example, to create a Range that represents letters from 'a' to 'z'?

Just like the IEnumerable interface, the implementation should be determined by the developer's own implementation.

For example, the Range for an array (which is the default) definitely represents non-negative indices, but when I use a custom Range, I can decide the start and end of the interval.

**Reply by huoyaoyuan** [reply, marked as answer]
> What I mean is that the semantics of Range are limited by the current implementation of the Index struct.
> 
> Instead of the usual semantics of Range representing an interval.

Both `System.Range` and the range expression mean "range of linear collection". They don't mean other ranges.
